### ISSUE TITLE
Proposal: Negate/Invert properties

### ISSUE THREAD

**Issue by ComradeVanti** [fscheck/FsCheck#614]
I am currently in the situation where I wrote a complex property and want to test that a specific value does not match it. I'm unsure if there is currently a way to do this. Here is an example of what I would have in mind:

```fsharp
let prop1 thing = ...

let prop2 thing = ...

let bothProps thing = (prop1 thing) .&. (prop2 thing) // True if both prop1 and prop2 are true

let notBothProps thing = Prop.not (bothProps thing) // Invert the property so it is true if either prop1 or prop2 fail
```

My suggestion is in the last line. I would love a `Prop.not`, `Prop.negate` or `Prop.invert` to flip properties. Could also add an operator such as `.!`.

Edit: If there is already a way to do this, I could not find it in the documentation. In that case, I request that it be documented somewhere.

Thanks

**Comment by ComradeVanti** [comment]
I just realized this would also allow us to create xor properties where one or the other, but not both must be true. Could also have an operator like `.|!.` or something.

**Comment by ComradeVanti** [comment]
@kurtschelfthout thanks for the reply, I get that you want to keep the API as simple as possible. Ill show you an example of how I would use property inversion and maybe cou can see the value of it, or tell me how to do this currently.

So lets say I have written my own generator for int-lists where each number is bigger than the previous one and I have written the following tests for it:

```fsharp
let isNotEmpty list = 
    not (list |> List.isEmpty) |@ "List is not empty"

let valuesAreOrdered list =
    list |> List.sort = list |@ "List is ordered"
    
let listIsValid list =
    (list |> isNotEmpty) .&. (list |> valuesAreOrdered)
```

Now lets say I also want to generate "invalid" lists, meaning they need to be either empty or not ordered. I could do this by writing inversions of all my existing properties:

```fsharp
let isEmpty list =
    list |> List.isEmpty |@ "List is empty"

let valuesAreNotOrdered list =
    list |> List.sort <> list |@ "List is not ordered"
    
let listIsInvalid list =
    (list |> isEmpty) .|. (list |> valuesAreNotOrdered)
```

And that works fine for this case, because there are only two properties to invert, but in the situation which made me write this issue my composite property is made up of like 10 smaller ones, which I would have to all invert indiviually and string together with `.|.`. What I would like instead is to just be able to go:

```fsharp
let listIsInvalid list =
    Prop.not (list |> listIsValid)
```

Lets say I use this property on the list `[1; 3; 2]`. In this case I would expect this property to succeed and if I use it on `[1; 2; 3]` I would expect it to fail. I you want I could also give an example for a XOR case, though I admit that one is pretty niche.

Hope this clears it up. If you could give me a hint on how you would solve this kind of situation usually, then I would also be very grateful.

**Comment by ploeh** [comment]
I'm not sure I understand this feature request.

It's entirely conceivable that you'll need to write properties that cover invalid inputs as well as valid inputs. Such properties, however, tend to be different from the properties that test happy paths.

Even for something like `isNotEmpty` or `valuesAreOrdered` you'd need a different generator to produce invalid inputs.

The provided examples look degenerate to me. Even if you could do what you suggest, what would you be testing? It seems that these properties express properties about a hypothetical FsCheck generator (_arbitrary_) rather than about an API or type.

I'm having trouble extrapolating from these examples to something that looks useful.

**Comment by ComradeVanti** [comment]
@ploeh it seems that maybe I have an incomplete or incorrect understanding of how to do pbt correctly. But here is an example. If I have a function that should act in a certain way for all valid inputs and a different way for all invalid inputs, I would have to create generators for all valid inputs and all invalid ones.

```fsharp
let ``True for valid inputs`` (ValidInput input) =
  input |> myFunc = true

let ``False for invalid inputs`` (InvalidInput input) =
  input |> myFunc = false
```

Now lets say I have written a generator for `ValidInput` with some rules that I have combined with `.&.` so that a `ValidInput` must satisfy all of them. Now if want to generate `InvalidInput` it would be neat to be able to simply invert the properties that I created for `ValidInput` and combine them with `.|.`.

I will close this issue soon, because again, I'm starting to think maybe I misunderstood some important pbt principle, but maybe someone could give me a hint of where my misconception is. Thanks a lot.

**Comment by ComradeVanti** [comment]
Seems, like what I'm trying to do is going into the direction of partition-testing, where I am basically saying "For partition A do X and for ¬A do Y".  I'm going to do some research into how to avoid this and how to do it "properly". With this, I'd be fine with closing the issue unless one of you has something to add.

**Comment by ploeh** [comment]
Here are a few thoughts. [Property-based testing isn't the same as partition testing](https://blog.ploeh.dk/2021/06/28/property-based-testing-is-not-the-same-as-partition-testing).

Even so, that doesn't mean that a feature like the proposed is inappropriate. I do, however, have some questions:

# Vacuous truths

Consider the FizzBuzz properties in the above article. There's one F# example:

```fsharp
[<Property(QuietOnSuccess = true)>]
let ``FizzBuzz.transform returns Buzz`` (number : int) =
    (number % 5 = 0 && number % 3 <> 0) ==> lazy
    let actual = FizzBuzz.transform number
    let expected = "Buzz"
    expected = actual
```

What's the inverted property of this property?

This property should hold for every `int`, so does this mean that for _no_ input, `expected` should _not_ be equal to `actual`. That would actually be a [vacuous truth](https://en.wikipedia.org/wiki/Vacuous_truth), so logically it still holds water. Is it useful, though?

Also, would this mean that _FsCheck_ needs some special-case logic to handle a scenario like this? It clearly isn't going to be able to produce 100 'inverse' inputs, because there are zero such input values.

As far as I can tell, you're going to run into similar issues with the canonical `List.rev` property example, which holds for all input.

# Contravariance

Another problem may present itself with contravariance. A unit test (and hence, a property) is essentially a [predicate, which we can regard as a contravariant functor](https://blog.ploeh.dk/2021/09/09/the-specification-contravariant-functor).

Thus, you might write a property like this (code may contain errors, I haven't tried to compile it):

```fsharp
[<Property>]
let myProperty (xs : Foo list) = myAssertion xs
```

What is the inverse property of this? How do you invert the input? If you use the [complement](https://en.wikipedia.org/wiki/Complement_(set_theory)) of `Foo list`, does that include `Bar list`, even if `Bar` is a subtype of `Foo`? And if you do, do you then expect the inverted assertion to pass?

# Complicated properties

While I agree that all code (and thereby also properties) should be as simple as possible, it's possible to write some quite complex properties. I'm [guilty of that myself](https://blog.ploeh.dk/2016/06/28/roman-numerals-via-property-based-tdd):

```fsharp
[<Property(QuietOnSuccess = true)>]
let ``symbols in descending order are additive`` () =
    let stringRepeat char count = String (char, count)
    let genSymbols count symbol =
        [0..count] |> List.map (stringRepeat symbol) |> Gen.elements
    let thousands =    genSymbols 3 'M'
    let fiveHundreds = genSymbols 1 'D'
    let hundreds =     genSymbols 3 'C'
    let fifties =      genSymbols 1 'L'
    let tens =         genSymbols 3 'X'
    let fives =        genSymbols 1 'V'
    let ones =         genSymbols 3 'I'
    let symbols = 
        [thousands; fiveHundreds; hundreds; fifties; tens; fives; ones]
        |> Gen.sequence
        |> Gen.map String.Concat
        |> Arb.fromGen
    Prop.forAll symbols <| fun s ->
 
        let actual = Numeral.tryParseRoman s
        
        let expected =
            s
            |> Seq.map (string >> Numeral.tryParseRoman)
            |> Seq.choose id
            |> Seq.sum
            |> Some            
        expected =! actual
```

Do we expect a framework like FsCheck to be able to invert a property like that one? Would it make sense?

# More than two pathways

Many algorithms or business rules imply more than two pathways through the code. Consider testing the Tennis point system.

When one player has forty points, there are three possible outcomes: The player wins, the score is _deuce_, or the other player's points are incremented. One of these outcomes [could be modelled like this](https://blog.ploeh.dk/2016/02/15/types-properties-software-properties-for-the-forties):

```fsharp
[<Property>]
let ``Given player: 40 when player wins then score is correct``
    (current : FortyData) =
 
    let actual = scoreWhenForty current current.Player
 
    let expected = Game current.Player
    expected =! actual
```

What's the inverse property here? Again, this property holds for all `FortyData` inputs, but even if you imagine that you could invert the scenario, is it interesting to assert that `actual` is _not_ equal to `Game current.Player`? Wouldn't it be more interesting to state what it's actually expected to be?

That's what a test like this one does:

```fsharp
[<Property>]
let ``Given player: 40 - other: 30 when other wins then score is correct``
    (current : FortyData) =
 
    let current = { current with OtherPlayerPoint = Thirty }
    let actual = scoreWhenForty current (other current.Player)
    Deuce =! actual
```

In short, is the inverse property really interesting for algorithms or rules that have more than two possible equivalence sets?

**Comment by jzabroski** [comment]
I don't think FsCheck should get into the business of "programming as inductive inference." 

If you're trying to negative/invert properties to do property-based testing, use a SAT Solver.

**Comment by ComradeVanti** [comment]
I have thought of one more example to illustrate what I mean, which I have also already encountered in my projects.

Let's say I have a constrained type `Age` which is essentially an `int` which is contrained to the range [0; 100]. I create a `tryMake` function for this type which looks like this:

```fsharp
let tryMake i =
  if i >= 0 && i <= 100 then Some (Age i)
  else None
```

Now I want to test this. Essentially there three partitions of inputs here. Too small, just right and too large.

Lets say I define my first test like this:

```fsharp
[<Property>]
let ``Some for valid inputs`` i =
  (i >= 0 && i <= 100) ==> lazy
  i |> Age.tryMake |> Option.isSome
```

Since this filter only matches a very small number of `ints`, I may choose to create a custom type and generator for valid inputs.

```fsharp
type Valid = Valid of int

let genValid =
  Gen.choose (0, 100)
  |> Gen.map Valid

// Create and register Arb

[<Property>]
let ``Some for valid inputs`` (Valid i) =
  i |> Age.tryMake |> Option.isSome
```

At this point I might also want to write tests for my `Valid` generator to make sure it matches my business logic. So maybe I would do something like this:

```fsharp

let isLargeEnough i =
  i >= 0 |@ "Is large enough"

let isSmallEnough i =
  i <= 100 |@ "Is small enough"

let isValid i =
   isLargeEnough i .&. isSmallEnough i

[<Property>]
let ``Valid gen is ok`` (Valid i) =
  isValid i
```

Now, time for the `None`tests. Again creating a custom generator (which I will not spell out here) which generates [... ;-1] or [101; ...] which I would want to use like this:

```fsharp
type Invalid = Invalid of int

let genInvalid =
  ...

// Create and register Arb

[<Property>]
let ``Some for invalid inputs`` (Invalid i) =
  i |> Age.tryMake |> Option.isNone
```

And now finally, I also want to create tests for my `Invalid` generator, which is finally where the inversion would come in. Instead of doing:

```fsharp

let isTooSmall i =
  i < 0 |@ "Is too small"

let isTooLarge i =
  i > 100 |@ "Is too large"

let isInvalid i =
  isTooSmall i .|. isTooLarge i

[<Property>]
let ``Invalid gen is ok`` (Invalid i) =
  isInvalid i
```

I would like to be able to just go:

```fsharp
[<Property>]
let ``Invalid gen is ok`` (Invalid i) =
  Prop.not (isValid i)
```

Hopefully now it's clear. Thanks for all your comments :)

**Comment by ploeh** [comment]
Would you also write properties that test that the properties that test your generators work correctly?

**Comment by ComradeVanti** [comment]
@ploeh I write tests for my API. However, in cases where in order to test my API I need to generate specific conditions (`ints` between 0 and 100) I add tests to make sure these conditions are met in my inputs. These generators are, in a way, part of my business logic definition, because they define what a `Valid` or `Invalid` input for my API is. As such I write tests for them to make sure they match my domain. 

I'm curious now as to how a situation like mine is "properly" done in PBT. Meaning, how would I test the following function (if at all, since the implementation is so trivial).

```fsharp
let tryMake i =
  if i >= 0 && i <= 100 then Some (Age i)
  else None
```

I realize I'm getting a little off-topic here, so if I should create a seperate issue, or discuss this elsewhere let me know.

**Comment by jzabroski** [comment]
@ComradeVanti I think @ploeh 's question flew over your head.

You're trying to solve your problem by creating a logical superstructure on your tests that your domain model doesn't have.  If the domain model shifts, you will have **no way** to accurately estimate the time to refactor the domain and fix all the tests.  And that's you, as the original author.  Now imagine some poor sap 5 years from now who has taken over your role. They'll probably just ask the business for new requirements and re-write everything.

**Comment by jzabroski** [comment]
I'll repeat my recommendation that if you want a model checker, use model checking software.

**Comment by ComradeVanti** [comment]
@jzabroski I'm not sure I understand what you mean, but I also feel that they way I'm currently doing it is off. I will look into model checking. With this, I think, we can close the issue. Thanks for all the feedback.

**Comment by jzabroski** [comment]
Alloy is a pretty friendly model checking language with decent community.  Amazon uses model checking for AWS APIs - you can read their experience report here: https://www.amazon.science/publications/model-checking-as-a-human-endeavor

**Comment by ploeh** [comment]
I apologise that the previous comment of mine was snarky.

It's sometimes appropriate to 'test your test code', but in general few people do it. It raises the question: _If you test your test code, then should you also test the test code that tests the test code?_

I wrote [an article](https://blog.ploeh.dk/2013/04/02/why-trust-tests) about that in 2013. It's not related to property-based testing, but applies as well.

I'm not going to claim that my way is the only appropriate way, but I've found that a major reason that I trust my tests is that I've seen them fail. This carries over to property-based testing as well.

---

Since you asked, though, here's how I would test `tryMake` (if I decided that it was warranted).

First write a property to verify successful creation:

```fsharp
[<Property>]
let ``Successful creation`` () =
    let validRange = Gen.choose (0, 100) |> Arb.fromGen
    Prop.forAll validRange <| fun i ->

        let actual = tryMake i

        actual |> Option.isSome
```

Following red-green-refactor, this implementation fails the test:

```fsharp
let tryMake i = None
```

Now make the test pass:

```fsharp
let tryMake i = Some (Age 0)
```

Now add a test for the failure case:

```fsharp
[<Property>]
let ``Failed creation`` () =
    let tooLow  = Gen.choose (Int32.MinValue, -1)
    let tooHigh = Gen.choose (101, Int32.MaxValue)
    let invalidAges = Gen.oneof [tooLow; tooHigh] |> Arb.fromGen
    Prop.forAll invalidAges <| fun i ->

        let actual = tryMake i

        Assert.Equal (None, actual)
```

This test fails until you make it pass:

```fsharp
let tryMake i =
    if 0 <= i && i <= 100
    then Some (Age 0)
    else None
```

It'd be fair to protest that this implementation is incorrect, but it passes all tests. As far as I can tell, it would also pass all the tests you suggest...

**Comment by jzabroski** [comment]
A related concern to `then Some (Age 0)` with "testing your tests" is _Gettier Problems_ in Justified-True-Belief theory.  Justified-True-Belief theory is a theory of how we know things, dating back to Plato and Socrates.  A Gettier Problem was a counter-example that showed how our justifications might be true but erroneous in the context of proving what we believe is knowledge.  This can happen in programming - imagine you test a "concurrent design" using linear, serial, deterministic inputs only.  The tests always pass, and so you believe you have correctly designed a concurrent system.  Yet, you overlook the fact that you _mocked_ a key detail that controls _non-deterministic choice_ to always return a certain value without delay.  Your system works correctly, but not for the reasons you've justified to yourself and codified in tests.

**Comment by ComradeVanti** [comment]
@ploeh yes I left out the test that checks if the `Age` actually contains the correct value, because I wanted to keep the example simple. Also in your example, you also define a partition for valid inputs expect you kept it local to the property instead of extracting it to a separate generator like I did.

The way I see it the line `let validRange = Gen.choose (0, 100) |> Arb.fromGen` essentially contains domain-defining logic because this line is what defines what a "valid" age is. As such I would add a test for this line to make sure it matches my domain requirements. For example, you actually made an error by making it `Gen.choose (0, 100)` instead of `Gen.choose (1, 100)` as defined by my domain. To make sure I catch such errors I would add a test for my generator which makes sure that "valid" inputs are [1; 100] and "invalid" ones are all `ints`excluding that range.

Idk if thats too much, but I always feel like I am leaving myself open to mistakes if I dont do it.

**Comment by ploeh** [comment]
@ComradeVanti To err is human. What makes it less likely that one makes an error like the one I did if you test your generators?

**Comment by ComradeVanti** [comment]
@ploeh I would add two tests which make sure that valid inputs are [1; 100] and invalid ones are anything excpet [1; 100]. I guess, its true that I could also mess up when writing these tests, but at the very least there is a well-defined place where to look for this error, as this would be more or less the only place where I define what makes a valid and invalid input.

**Comment by ploeh** [comment]
TBH, to test something like `Age`, I think it'd be easier and safer to write two parametrised tests with a few examples:

```fsharp
[<Theory>]
[<InlineData   1>]
[<InlineData   2>]
[<InlineData  50>]
[<InlineData  99>]
[<InlineData 100>]
let ``Successful creation, examples`` i =
    let actual = tryMake i
    Assert.True actual.IsSome

[<Theory>]
[<InlineData   0>]
[<InlineData  -1>]
[<InlineData  -9>]
[<InlineData 101>]
[<InlineData 200>]
let ``Failed creation, examples`` i =
    let actual = tryMake i
    Assert.True actual.IsNone
```

It's much easier to catch off-by-one errors like mine in a code review when you have some concrete examples to consider.

Property-based testing is great, but beware that it doesn't become a [Golden Hammer](https://en.wikipedia.org/wiki/Law_of_the_instrument).

**Comment by ComradeVanti** [comment]
Yeah I noticed myself going into that direction of attempting to use it on everything. I guess the next level of my testing-journey will be to learn on when the different styles of tests are besrt applied. Thanks a lot for the advice.

**Comment by jzabroski** [comment]
I agree with Mark.

I was going to explain that having a continuously valid range of numbers is easy to write a data generator for, but it gets annoying when you have certain values in the middle that are not allowed.  An example is synthetic primary keys where 4 IDs in the range from 1 to 100 are "special": 91, 92, 93, 94.  I had such a model to deal with two weeks ago.

Additionally, I would reify your data validator in the domain model, probably with a library like FluentValidation.  I think there is a F#-specific validation library but I cant remember the name of it.

**Comment by bartelink** [comment]
>Additionally, I would reify your data validator in the domain model, probably with a library like FluentValidation. I think there is a F#-specific validation library but I cant remember the name of it.

https://github.com/pimbrouwers/Validus
